# Melakukan Pemrosesan data menggunakan Python


## import library pandas pada py

In [ ]:
import pandas as pd

## membaca file CSV

In [ ]:
df=pd.read_csv('/content/data_kotor_alat_berat_b2b.csv')

### Melakukan proses pencarian outlier pada data frame

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3050 entries, 0 to 3049
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Order_ID              3050 non-null   object
 1   Tanggal_Transaksi     3050 non-null   object
 2   Region                3004 non-null   object
 3   Sektor_Industri       3020 non-null   object
 4   Kategori_Produk       3050 non-null   object
 5   Kuantitas             3050 non-null   int64 
 6   Harga_Satuan_IDR      3050 non-null   int64 
 7   Total_Nilai_Deal_IDR  3050 non-null   int64 
 8   Status_Transaksi      3050 non-null   object
 9   Revenue_IDR           3050 non-null   int64 
dtypes: int64(4), object(6)
memory usage: 238.4+ KB


#### Hasil Analisis pertama<br>
1.terlihat bahwa pada range index kita ada 3050 total namun untuk setiap atribut kita ada beberapa kolom yang memiliki data kurang dari 3050 yang menandakan bahwa ada data yang missing value<br>
2.beberapa kolom seperti tanggal_transaksi masih memliki dtype yang salah jadi harus kita ubah menggunakan time series

In [ ]:
df.duplicated().sum()#mencari duplikasi data

np.int64(50)

#### Hasil Analisis Kedua<br>
Terlihat bahwa ada 50 data yang benar benar kembar kemungkinan ada double click input data

In [ ]:
df.describe()

,Kuantitas,Harga_Satuan_IDR,Total_Nilai_Deal_IDR,Revenue_IDR
count,3050.000000,3.050000e+03,3.050000e+03,3.050000e+03
mean,2.640984,1.012788e+09,2.204380e+09,1.339242e+09
std,1.457823,8.247002e+08,1.714601e+09,1.689269e+09
min,-2.000000,0.000000e+00,-5.911967e+09,0.000000e+00
25%,1.000000,3.474591e+08,8.889394e+08,0.000000e+00
50%,2.000000,5.771722e+08,1.744149e+09,6.612965e+08
75%,4.000000,1.489585e+09,3.151125e+09,2.133589e+09
max,5.000000,2.999149e+09,7.491271e+09,7.491271e+09


#### Hasil Analisis Kedua<br>
1.terlihat bahwa ada kuantitas yang min dimana kuantitas tidak mungkin mines karna tidak mungkin seseorang membli alat yang mines yang mana menghasilkan harga dealnya mencapai mines 5M<br>
2.ada harga barang yang 0 dimana tidak mungkin kita menjual produk dengan harga o rupiah

### proses pencarian outlier,missing value dan juga duplikasi data telah cukup sekarang kita masuk ke dalam pembersihan datanya

#### menghilangakn dupilkasi data karna murni 100% itu sudah sangat besar kemungkinan human eror jadi solusi kita akan melakukan syntax drop.dupilcate saja


In [ ]:
df = df.drop_duplicates()

In [ ]:
df.duplicated().sum()

np.int64(0)

##### Sudah terlihat bahwa duplikasi data sudah 0

In [ ]:
df.columns = df.columns.str.lower()#mengubah header column jadi huruf kecil
for col in df.select_dtypes(include='object').columns:
  df[col]=df[col].str.lower()#untuk mempermudah kita rubah tulisan tipe data objek jadi huruf kecil semua menggunakan for looping

In [ ]:
df=df[(df['kuantitas']>0) & (df['harga_satuan_idr']>0)]

In [ ]:
df.describe()

,kuantitas,harga_satuan_idr,total_nilai_deal_idr,revenue_idr
count,2975.000000,2.975000e+03,2.975000e+03,2.975000e+03
mean,2.666218,1.015103e+09,2.235213e+09,1.333806e+09
std,1.423404,8.228019e+08,1.686237e+09,1.689089e+09
min,1.000000,1.002952e+08,1.003282e+08,0.000000e+00
25%,1.000000,3.505701e+08,9.115369e+08,0.000000e+00
50%,2.000000,5.798388e+08,1.761490e+09,6.484743e+08
75%,4.000000,1.490830e+09,3.179588e+09,2.120992e+09
max,5.000000,2.999149e+09,7.491271e+09,7.491271e+09


##### oke tahap selanjutnya kita akan memperbaiki mis value nya

In [ ]:
# 2. Isi data kosong (Missing Value) agar data penjualannya tidak terbuang
df['region'] = df['region'].fillna('unknown')
df['sektor_industri'] = df['sektor_industri'].fillna('lainnya')

/tmp/ipykernel_2120/734405703.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['region'] = df['region'].fillna('unknown')
/tmp/ipykernel_2120/734405703.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sektor_industri'] = df['sektor_industri'].fillna('lainnya')


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2975 entries, 0 to 3049
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   order_id              2975 non-null   object
 1   tanggal_transaksi     2975 non-null   object
 2   region                2975 non-null   object
 3   sektor_industri       2975 non-null   object
 4   kategori_produk       2975 non-null   object
 5   kuantitas             2975 non-null   int64 
 6   harga_satuan_idr      2975 non-null   int64 
 7   total_nilai_deal_idr  2975 non-null   int64 
 8   status_transaksi      2975 non-null   object
 9   revenue_idr           2975 non-null   int64 
dtypes: int64(4), object(6)
memory usage: 255.7+ KB


In [ ]:
df.isnull().sum()

,0
order_id,0
tanggal_transaksi,0
region,0
sektor_industri,0
kategori_produk,0
kuantitas,0
harga_satuan_idr,0
total_nilai_deal_idr,0
status_transaksi,0
revenue_idr,0


##### dari dua info diatas telah di simpulkan bahwa missing value juga sudah kita perbaiki

#### sekarang kita akan coba mengubah tipe data yang tidak sesuai

In [ ]:
# 3. Ubah kolom tanggal dari teks (object) menjadi format waktu (datetime)
df['tanggal_transaksi'] = pd.to_datetime(df['tanggal_transaksi'])

/tmp/ipykernel_2120/479757459.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['tanggal_transaksi'] = pd.to_datetime(df['tanggal_transaksi'])


#### oke semua otlier missing value begitu juga duplikasi data sudah kita selesaikan untuk saya biasanya langkah paling akhir yang selalu saya lakukan adalah pengecekan typo

In [ ]:
for col in ['region', 'sektor_industri', 'kategori_produk', 'status_transaksi']:
    print(df[col].value_counts())
    print("-" * 30)

region
sumatera      614
jawa          599
papua         588
sulawesi      567
kalimantan    562
unknown        45
Name: count, dtype: int64
------------------------------
sektor_industri
pertambangan         942
perkebunan           657
konstruksi           541
manufaktur           431
logistik & gudang    374
lainnya               30
Name: count, dtype: int64
------------------------------
kategori_produk
genset (generator set)    772
excavator                 730
forklift                  729
tractor agrikultur        700
forklip                    25
ekskavator                 19
Name: count, dtype: int64
------------------------------
status_transaksi
won (berhasil)    1778
lost (gagal)       890
in progress        307
Name: count, dtype: int64
------------------------------


##### bisa kita liat ada beberpa yang missing value seperti forklip ekskavator

In [ ]:
df['kategori_produk'] = df['kategori_produk'].replace({
    'ekskavator': 'excavator',
    'forklip': 'forklift'
})

/tmp/ipykernel_2120/2563569877.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['kategori_produk'] = df['kategori_produk'].replace({


In [ ]:
for col in ['region', 'sektor_industri', 'kategori_produk', 'status_transaksi']:
    print(df[col].value_counts())
    print("-" * 30)

region
sumatera      614
jawa          599
papua         588
sulawesi      567
kalimantan    562
unknown        45
Name: count, dtype: int64
------------------------------
sektor_industri
pertambangan         942
perkebunan           657
konstruksi           541
manufaktur           431
logistik & gudang    374
lainnya               30
Name: count, dtype: int64
------------------------------
kategori_produk
genset (generator set)    772
forklift                  754
excavator                 749
tractor agrikultur        700
Name: count, dtype: int64
------------------------------
status_transaksi
won (berhasil)    1778
lost (gagal)       890
in progress        307
Name: count, dtype: int64
------------------------------


In [ ]:
df.head()

,order_id,tanggal_transaksi,region,sektor_industri,kategori_produk,kuantitas,harga_satuan_idr,total_nilai_deal_idr,status_transaksi,revenue_idr
0,trx-21517,2022-12-15,sumatera,perkebunan,excavator,1,2277284162,2277284162,lost (gagal),0
1,trx-22369,2021-09-24,kalimantan,manufaktur,forklift,5,409489290,2047446450,won (berhasil),2047446450
2,trx-21961,2022-07-17,sumatera,pertambangan,forklift,1,273735241,273735241,lost (gagal),0
3,trx-20343,2021-12-29,sulawesi,manufaktur,genset (generator set),4,367795935,1471183740,lost (gagal),0
4,trx-22663,2021-01-21,sulawesi,perkebunan,forklift,2,288704212,577408424,lost (gagal),0


In [ ]:
df.tail()

,order_id,tanggal_transaksi,region,sektor_industri,kategori_produk,kuantitas,harga_satuan_idr,total_nilai_deal_idr,status_transaksi,revenue_idr
3045,trx-21638,2023-03-20,kalimantan,perkebunan,excavator,2,1530294934,3060589868,won (berhasil),3060589868
3046,trx-21095,2021-10-10,papua,pertambangan,tractor agrikultur,5,1164982741,5824913705,won (berhasil),5824913705
3047,trx-21130,2022-02-23,sulawesi,pertambangan,forklift,1,428099624,428099624,lost (gagal),0
3048,trx-21294,2023-12-05,unknown,manufaktur,excavator,1,1806150281,1806150281,won (berhasil),1806150281
3049,trx-20860,2023-06-20,sulawesi,perkebunan,genset (generator set),3,236879888,710639664,won (berhasil),710639664


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2975 entries, 0 to 3049
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   order_id              2975 non-null   object        
 1   tanggal_transaksi     2975 non-null   datetime64[ns]
 2   region                2975 non-null   object        
 3   sektor_industri       2975 non-null   object        
 4   kategori_produk       2975 non-null   object        
 5   kuantitas             2975 non-null   int64         
 6   harga_satuan_idr      2975 non-null   int64         
 7   total_nilai_deal_idr  2975 non-null   int64         
 8   status_transaksi      2975 non-null   object        
 9   revenue_idr           2975 non-null   int64         
dtypes: datetime64[ns](1), int64(4), object(5)
memory usage: 255.7+ KB


In [ ]:
df.describe()

,tanggal_transaksi,kuantitas,harga_satuan_idr,total_nilai_deal_idr,revenue_idr
count,2975,2975.000000,2.975000e+03,2.975000e+03,2.975000e+03
mean,2022-06-27 08:31:08.369747968,2.666218,1.015103e+09,2.235213e+09,1.333806e+09
min,2021-01-01 00:00:00,1.000000,1.002952e+08,1.003282e+08,0.000000e+00
25%,2021-09-19 00:00:00,1.000000,3.505701e+08,9.115369e+08,0.000000e+00
50%,2022-07-05 00:00:00,2.000000,5.798388e+08,1.761490e+09,6.484743e+08
75%,2023-03-30 12:00:00,4.000000,1.490830e+09,3.179588e+09,2.120992e+09
max,2024-01-01 00:00:00,5.000000,2.999149e+09,7.491271e+09,7.491271e+09
std,NaN,1.423404,8.228019e+08,1.686237e+09,1.689089e+09


In [ ]:
### Oke proses cleansing pada data kita sudah selesai sekarang kita akan masuk untuk membuat dataset setalah pemrosesan

In [ ]:
df.to_csv('data_transaksi_bersih.csv', index=False)